Examples/Algorithms/Geant4HepMC/src/PrimaryGeneratorAction.cpp

```diff
@@ -0,0 +1,68 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "PrimaryGeneratorAction.hpp"
+#include <stdexcept>
+#include <G4Event.hh>
+#include <G4ParticleDefinition.hh>
+#include <G4ParticleGun.hh>
+#include <G4RandomDirection.hh>
+#include <G4UnitsTable.hh>
+#include <Randomize.hh>
+
+ActsExamples::PrimaryGeneratorAction*
+    ActsExamples::PrimaryGeneratorAction::s_instance = nullptr;
+
+ActsExamples::PrimaryGeneratorAction::PrimaryGeneratorAction(G4int randomSeed1,
+                                                             G4int randomSeed2)
+    : G4VUserPrimaryGeneratorAction(), m_particleGun(nullptr) {
+  // Configure the run
+  if (s_instance) {
+    throw std::logic_error("Attempted to duplicate a singleton");
+  } else {
+    s_instance = this;
+  }
+
+  // Configure the gun
+  G4int nofParticles = 1;
+  m_particleGun = std::make_unique<G4ParticleGun>(nofParticles);
+
+  // Prepare the particle table
+  m_particleTable = G4ParticleTable::GetParticleTable();
+
+  // Set the random seeds
+  CLHEP::HepRandom::getTheEngine()->setSeed(randomSeed1, randomSeed2);
+}
+
+ActsExamples::PrimaryGeneratorAction::~PrimaryGeneratorAction() {
+  s_instance = nullptr;
+}
+
+ActsExamples::PrimaryGeneratorAction*
+ActsExamples::PrimaryGeneratorAction::instance() {
+  // Static acces function via G4RunManager
+  return s_instance;
+}
+
+void ActsExamples::PrimaryGeneratorAction::prepareParticleGun(
+    const ActsExamples::SimParticle& part) {
+  // Particle type
+  G4ParticleDefinition* particle = m_particleTable->FindParticle(part.pdg());
+  m_particleGun->SetParticleDefinition(particle);
+  // Particle properties
+  const auto pos = part.position();
+  const auto dir = part.unitDirection();
+  m_particleGun->SetParticlePosition({pos[0], pos[1], pos[2]});
+  m_particleGun->SetParticleMomentum(part.absMomentum());
+  m_particleGun->SetParticleMomentumDirection({dir[0], dir[1], dir[2]});
```

Comment:
According to the [Geant Docs](https://geant4-userdoc.web.cern.ch/geant4-userdoc/UsersGuides/ForApplicationDeveloper/BackupVersions/V10.6/html/GettingStarted/eventDef.html#public-methods-of-g4particlegun) the particle gun should have a ` SetParticleTime` method. Maybe it is only available in newer versions?

Reply:
It looks like it's for 4.10.06. We support currently only older versions.

Reply:
Understood.